server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileResource.java

```diff
@@ -535,6 +535,33 @@ public boolean removeResourceFile() {
     return true;
   }
 
+  /**
+   * Move its data file, resource file, and modification file physically.
+   *
+   * @return migrated data file
+   */
+  public File migrate(File targetDir) {
+    // get the resource and mod files
+    File resourceFile = fsFactory.getFile(file.getPath() + RESOURCE_SUFFIX);
+    File modFile = fsFactory.getFile(file.getPath() + ModificationFile.FILE_SUFFIX);
+
+    // get the target file locations
+    File migratedFile = fsFactory.getFile(targetDir, file.getName());
+    File migratedResourceFile = fsFactory.getFile(targetDir, file.getName() + RESOURCE_SUFFIX);
+    File migratedModificationFile =
+        fsFactory.getFile(targetDir, file.getName() + ModificationFile.FILE_SUFFIX);
+
+    // move
+    fsFactory.moveFile(file, migratedFile);
+    if (resourceFile.exists()) {
+      fsFactory.moveFile(resourceFile, migratedResourceFile);
+    }
+    if (modFile.exists()) {
+      fsFactory.moveFile(modFile, migratedModificationFile);
+    }
+    return migratedFile;
+  }
```

Comment:
This method isn't atomic. If the system crashes or exception occurs when this method is running, some files may still in the origin data directory, not in the target directory. How to recover our system from this scenario? Please think of a solution to this problem.

Reply:
Fixed by adding MigratingFileLogManager that keeps track of file migrations, and provides the ability to recover.